llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp

```diff
@@ -12811,11 +12930,24 @@ unsigned ARMAsmParser::validateTargetOperandClass(MCParsedAsmOperand &AsmOp,
     if (hasV8Ops() && Op.isReg() && Op.getReg() == ARM::SP)
       return Match_Success;
     return Match_rGPR;
-  case MCK_GPRPair:
-    if (Op.isReg() &&
-        MRI->getRegClass(ARM::GPRRegClassID).contains(Op.getReg()))
+  // If trying to match a VecListDPair with a Q register, convert Q to list
+  case MCK_VecListDPair:
+    if (Op.isQReg() && !hasMVE()) {
+      auto DPair = getDRegFromQReg(Op.getReg());
+      DPair = MRI->getMatchingSuperReg(
+          DPair, ARM::dsub_0, &ARMMCRegisterClasses[ARM::DPairRegClassID]);
+      Op.setVecListDPair(DPair);
```

Comment:
This method shouldn't mutate the operand.
You may want to modify the `ARMOperand::isVecListDPair` instead to accept single register and `ARMOPerand::addXXXOperand` to render the register as a list.

Reply:
Ah this makes sense, I didn't know where to do it. I followed the pattern of llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:1305 which does mutation at this point.
The functions needed for this conversion comes from the `MCRegisterInfo` object `MRI` which is a member of the `ARMAsmParser` class. This isn't available for at `ARMOperand::addXXXOperand` so this is quite challenging. I would like to avoid finding a way of plumbing this through if possible (I guess `ARMOperand` would need a reference to its parser which feels a bit gross) do you think it is wrong if this stays here or is it a matter of preference?

Reply:
> The functions needed for this conversion comes from the MCRegisterInfo object MRI which is a member of the ARMAsmParser class. This isn't available for at ARMOperand::addXXXOperand so this is quite challenging. 

This is a known issue, and you correctly guessed one way to overcome it. It would be better if addXXXOperand methods were members of AsmParser instance, where everything that is needed is available, but they are not, and it is difficult to change now.

> do you think it is wrong if this stays here or is it a matter of preference?

I guess it can stay here, but "mutate it at your own risc". Once the operand is mutated, it can no longer be matched as non-mutated.

Reply:
Ah the mutation not being reversible for future matching is a good point, I hadn't thought about that. I think in this case its okay as the circumstances where this conversion would happen are very specific but I will add a comment explaining at the very least.